Add `migrations has-pending-model-changes` Command

justindbaur opened the pull request:
<!--
Please check whether the PR fulfills these requirements
-->

Fixes #26348

- [x] I've read the guidelines for [contributing](CONTRIBUTING.md) and seen the [walkthrough](https://youtu.be/9OMxy1wal1s?t=1869)
- [x] I've posted a comment on an issue with a detailed description of how I am planning to contribute and got approval from a member of the team
- [x] The code builds and tests pass locally (also verified by our automated build checks)
- [x] Commit messages follow this format:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Code follows the same patterns and style as existing code in this repo

I could not find anything that tests `MigrationOperations` methods or anything that directly tests the CLI. Please let me know if I missed anything and I'd be happy to write some tests.

My biggest question is the code I added in `MigrationsOperations`. It seems like a bit to much code for what normally goes in those methods but I don't know where the ideal location for this logic would sit. The code I wrote is a combination of the existing [code here](https://github.com/dotnet/efcore/blob/81c2440d848ecfe2ad9b8147150d607526d036b8/src/EFCore.Design/Migrations/Design/MigrationsScaffolder.cs#L143-L149) and from the comment on the [issue here](https://github.com/dotnet/efcore/issues/26348#issuecomment-1243730207).

Also feel free to let me know the exact wording you would like for the resource strings, I doubt what I have there is great.

bricelam commented on src/EFCore.Design/Design/Internal/MigrationsOperations.cs:
```diff
@@ -242,6 +243,36 @@ public class MigrationsOperations
+
+        var model = scope.ServiceProvider.GetRequiredService<IModel>();
+
+        var hasDifferences = modelDiffer.HasDifferences(lastModel, model.GetRelationalModel());
```
@ajcvickers This checks pending model changes

bricelam commented on src/ef/Commands/MigrationsCommand.cs:
```diff
@@ -17,6 +17,7 @@ public override void Configure(CommandLineApplication command)
         command.Command("list", new MigrationsListCommand().Configure);
         command.Command("remove", new MigrationsRemoveCommand().Configure);
         command.Command("script", new MigrationsScriptCommand().Configure);
+        command.Command("check-pending", new MigrationsCheckPendingCommand().Configure);
```
We need to discuss the name in a design meeting. This proposal was intended to check for any un-applied migrations.

bricelam replied:
See [design meeting notes](https://github.com/dotnet/efcore/pull/31164#issuecomment-1629739850)

bricelam commented:
Open questions:

- Should we implement context.Database.HasChangesForAddMigration() at the same time and have it call that? (tentative method name)
- What should this output? Should it summarize the changes pending? Would a --dry-run flag on migrations add be better?
-should this have a PMC equivalent?
-Should it support --json?

bricelam commented on src/EFCore.Design/Properties/DesignStrings.resx:
```diff
@@ -350,6 +350,9 @@ Change your target project to the migrations project by using the Package Manage
     <value>Database '{name}' did not exist, no action was taken.</value>
   </data>
+  <data name="PendingMigrations" xml:space="preserve">
+    <value>You have pending migrations.</value>
```
Maybe something like:

> Changes have been made to the model since the last migration. Add a new migration.

bricelam commented on src/dotnet-ef/Properties/Resources.resx:
```diff
@@ -225,6 +225,9 @@
     <value>The target runtime to bundle for.</value>
   </data>
+  <data name="MigrationsCheckPendingDescription" xml:space="preserve">
+    <value>Checks if there are pending migrations.</value>
```
Maybe:

> Checks if any changes have been made to the model since the last migration.

bricelam replied:
(Don't forget to update the resources in `ef` too.)

bricelam commented on src/ef/Commands/MigrationsCheckPendingCommand.cs:
```diff
@@ -0,0 +1,16 @@
+
+        executor.CheckPendingMigrations(Context!.Value());
+
+        return base.Execute(args);
```
We want this to return an error code when there are pending migrations so this can easily be utilized in a CI.

bricelam replied:
Oh I see, you're relying on an exception to be thrown. That's probably fine until we decide whether to add a `--json`

bricelam commented on src/ef/Commands/MigrationsCheckPendingCommand.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        using var executor = CreateExecutor(args);
+
+        executor.CheckPendingMigrations(Context!.Value());
```
We should probably output something when there are no changes. Something like:

> No changes have been made to the model since the last migration.

bricelam commented:
You can add tests to [Tests/Design/Internal/MigrationsOperationsTest.cs](https://github.com/dotnet/efcore/blob/main/test/EFCore.Design.Tests/Design/Internal/MigrationsOperationsTest.cs). See the ones in [DbContextOperationsTest](https://github.com/dotnet/efcore/blob/main/test/EFCore.Design.Tests/Design/Internal/DbContextOperationsTest.cs) for a better example.

bricelam commented:
## 📝 Design Meeting Notes

- The command should be called `dotnet ef migrations has-pending-model-changes`
- The logic for this should be implemented as an extension method in [RelationalDatabaseFacadeExtensions](https://github.com/dotnet/efcore/blob/main/src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs) called `HasPendingModelChanges` that returns a `bool`
- A corresponding PMC command, and the --json option should wait for more/actual user requests before implementing them
- The command should always output some message saying whether or not there are changes (as already requested in PR review)
  - A summary of changes is not useful/actionable (just run migrations add)

bricelam commented:
@justindbaur Alright, the comments should contain all the changes we'd like to see before mering this. Let us know if you have any follow up questions. It's looking really good.

justindbaur commented:
@bricelam Thank you so much, I did start trying to write some tests but was having some issues because `MockAssembly.Name != typeof(TestContext). Assembly.Name` but maybe with making it an extension method off `DatabaseFacade` makes writing a test a bit easier.  But thanks for the feedback, I'll hopefully have something ready for a second pass in the next couple days.

justindbaur commented:
Hi @bricelam I believe I have addressed all feedback, the tests are not the prettiest things I have ever written. I had a bit of a learning curve trying to figure out how a snapshot model and design time model are created so I could best fake them while still achieving a high probability that the test is actually testing something. Let me know if you have a better way of achieving what I did.

bricelam commented on src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs:
```diff
@@ -952,6 +952,39 @@ public static bool IsRelational(this DatabaseFacade databaseFacade)
+    /// <summary>
+    ///     Returns <see langword="true" /> if the model has pending changes to be applied.
+    /// </summary>
+    /// <param name="databaseFacade">Tbe facade from <see cref="DbContext.Database"/>.</param>
```
```suggestion
    /// <param name="databaseFacade">The facade from <see cref="DbContext.Database"/>.</param>
```

bricelam reviewed (commented):
LGTM

justindbaur commented:
@bricelam Thank you so much for taking it across the finish line. Can't wait to add this to our CI!
